stackrox/stackrox
# ROX-14693: reconcile(create/update operation) declarative resources

## Description

Within this PR we create/update declarative resources found within the declarative configuration during reconciliation.
Deletion of declarative resources will be handled in a separate ticket.

The approach taken within this PR is to reconcile each resource individually. Reconciliation is done in order from referenced resources to referring resources(for example, access scope is referenced by role, so access scopes are reconciled first). That allows to avoid resolving references as it is guaranteed that in a valid configuration referenced resources will be reconciled first. 
It is possible that some state of configuration will not be valid(for example, if referenced and referring resources are located in different files), but important thing is that the configuration state in the backend will *eventually* catch on to changes within the files.

**A few other decisions made in this PR:**
1. The alternative to writing `reconcile*` functions in specific order was to introduce `protoTypeUpsertOrder` list of proto types. However, the problem was to generalize the type of reconcile function because of generic limitations in Golang.
2. `ReconciliationErrorReporter` interface decided to be made similar to `integrationHealthReporter`(every resource health and errors are treated separately). However, this is not principal part of this PR and can be changed later
3.  For auth providers `loginURL` field is filled within `registry.CreateProvider` call, `enabled` and `active` is set by default to true during transformation. There is no special handling of `basic` auth providers since declarative auth provider can't be of `basic` type.
4. Every datastore(except auth provider) got `Upsert` function. An alternative to supporting this function would be to execute `Get*` and then conditionally `Update*` or `Add*` datastore methods. There are two reasons not to follow the alternative approach: saving extra `Get*` call and not delegating datastore responsibilities to the datastore client.

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

**Testing plan(TBE - to be executed)**:
1. Manual test(configure declarative config - see if it's being reconciled)
2. Datastore level tests


Manually testing declarative configurations:
1. Create the helm chart:
```bash
roxctl helm output central-services --image-defaults=opensource
```
2. Define the following `values.yaml` file:
```yaml
central:
  declarativeConfiguration:
    mounts:
      configMaps:
        - permission-set
        - access-scope
        - auth-provider
        - role
customize:
  envVars:
    ROX_DECLARATIVE_CONFIGURATION: true
```
3. Install the central services chart:
```bash
helm install stackrox-central-services -n stackrox -f ./values.yaml ./stackrox-central-services-chart
```
4. Verify that central is running and the `ROX_DECLARATIVE_CONFIGURATION` variable is set. The mount points for declarative configuration should exist as well.
5. Create the first config map containing a permission set:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: permission-set
  namespace: stackrox
data:
  permission-set: |
    name: hello-world
    description: the first declaratively created permission set
    resources:
    - resource: Integration
      access: READ_ACCESS
    - resource: Administration
      access: READ_ACCESS
    - resource: Access
      access: READ_ACCESS
EOF
```
6. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the permission set `hello-world` exists.
7. Create the second config map containing an access scope:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: access-scope
  namespace: stackrox
data:
  access-scope: |
    name: hello-world
    description: the first declaratively created access scope
    rules:
      included:
      - cluster: remote
EOF
```
8. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the access scope `hello-world` exists.
9. Create the third config map containing a role:
```bash
kubectl apply -f - <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: role
  namespace: stackrox
data:
  role: |
    name: hello-world
    description: the first declaratively created role
    permissionSet: hello-world
    accessScope: hello-world
EOF
```
10. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the role`hello-world` exists.
11. Create the fourth config map containing an auth provider:
```bash
apiVersion: v1
kind: ConfigMap
metadata:
  name: auth-provider
  namespace: stackrox
data:
  auth-provider: |
    name: hello-world
    minimumRole: "None"
    uiEndpoint: localhost:8000
    groups:
    - key: "email"
      value: "<redacted>"
      role: "Admin"
    oidc:
      issuer: <redacted>
      mode: fragment
      clientID: <redacted>

EOF
```
12. Verify within the a) logs of central that no error occurred during reconciliation b) within the UI / API the auth provider`hello-world` exists and you are able to successfully login, and group mapping works.
